### PR TITLE
cl/beacon/synced_data: swap head state instead of copying it

### PR DESCRIPTION
**Pull request by awskii**
After each FCU, synced-data took two full `CachingBeaconState` copies under the write lock, mostly rebuilding the public-key index. Aggregate gossip validation waited on that lock. This PR copies once into the spare state, then swaps pointers after the copy succeeds; a failed copy keeps the last valid head.

| metric | base | patched |
|---|---:|---:|
| write-lock hold p50 | 470 ms | 240 ms |
| blocked aggregate validations, max wait | 29, 473 ms | 3, 61 ms |
| `Post forkchoice operations` p50 / p90 | 480 / 795 ms, 357 slots | 251 / 560 ms, 304 slots |

arb1-dev, mainnet archive at tip; base fb7c9898b96 / patched base + df0f3e3afe8; traces 05:07-05:13Z / 05:24-05:30Z, logs 04:00:00-05:17:47Z / 05:20:00-06:26:42Z.

### CODE REVIEW

File: cl/beacon/synced_data/synced_data.go [hunk 1]
```diff
@@ -94,14 +94,7 @@ func (s *SyncedDataManager) OnHeadState(newState *state.CachingBeaconState) (err
...
 	if s.headState != nil {
-		if s.previousHeadState != nil {
-			err = s.headState.CopyInto(s.previousHeadState)
-		} else {
-			s.previousHeadState, err = s.headState.Copy()
-		}
-		if err != nil {
-			return err
-		}
+		s.previousHeadState, s.headState = s.headState, s.previousHeadState
```
[the remark this answers is not in the document]
Comment (awskii) [reply cl/beacon/synced_data/synced_data.go]: Fixed in 06c235c8d7d. Both methods now copy into the spare state, or a fresh `Copy()` when none exists, and swap and store the identity only after the copy succeeds. A failed update keeps the last valid head. `main` had the half-copied head case too, because its `CopyInto` writes the head in place. `TestFailedHeadUpdateKeepsHead` covers the no-spare and spare paths.

File: cl/beacon/synced_data/synced_data.go [hunk 2]
```diff
@@ -92,33 +92,20 @@ func (s *SyncedDataManager) OnHeadState(newState *state.CachingBeaconState) (err
...
-	// Update headState with the new state.
-	if s.headState == nil {
-		s.headState, err = newState.Copy()
+	next := s.previousHeadState
+	if next == nil {
+		next, err = newState.Copy()
 	} else {
-		err = newState.CopyInto(s.headState)
+		err = newState.CopyInto(next)
 	}
```
[the remark this answers is not in the document]
Comment (awskii) [reply cl/beacon/synced_data/synced_data.go]: Fixed in af80fececeb. A failed copy now sets `previousHeadState` to nil, so `ViewPreviousHeadState` returns `ErrPreviousStateNotAvailable`. The fork-graph cache treats that as a miss and reads the state from disk. `OnHeadState` now computes `BlockRoot()` before delegating, so a root error returns before anything is written. `TestFailedHeadUpdateKeepsHead` checks that previous is dropped.

File: cl/beacon/synced_data/synced_data.go [hunk 3]
```diff
@@ -82,45 +82,12 @@ func (s *SyncedDataManager) SelectedHead() (common.Hash, uint64, bool) {
...
-		s.headState, err = newState.Copy()
-	} else {
-		err = newState.CopyInto(s.headState)
-	}
-	if err != nil {
-		return err
-	}
-	blkRoot, err = newState.BlockRoot()
+func (s *SyncedDataManager) OnHeadState(newState *state.CachingBeaconState) error {
+	blkRoot, err := newState.BlockRoot()
```
[the remark this answers is not in the document]
Comment (awskii) [reply cl/beacon/synced_data/synced_data.go:89]: Fixed in 76158a323c2. `OnHeadState` checks `enabled` first again, so a disabled manager returns before touching `newState`, as it did before the delegation. `TestDisabledHeadUpdateIsNoop` calls it with a nil state on a disabled manager.

**Review by domiwei — COMMENTED** [review]
Reviewed at b9b1d878e7. The spare-first copy and swap implementation looks correct, but I am holding approval for two missing tests on the changed ownership boundary:

1. Please add deterministic real OnHeadStateWithBlockRoot interleavings with ViewHeadState and ViewPreviousHeadState, plus UnsetHeadState, in both reader-to-writer and writer-to-reader order. Assert the writer cannot recycle either buffer while a callback owns it, and that ViewHeadStateWithIdentity, StateHead, and the scalar getters match the published state afterward. The existing concurrency test writes manager internals directly, so it does not exercise this path.

2. Please cover the authoritative prior-root integration: publish a state whose nonzero previousStateRoot is intentionally distinct from ordinary recomputation, publish the next head, then resolve the first root through the fork-graph previous-head fast path. The current empty Phase0 states only test recomputable roots; this path matters around Gloas and deferred payload handling.

The current tests do cover first, warm, and reused buffers plus fresh and recycled CopyInto failures. On failure the partial spare is discarded while the current head remains intact; please also assert its atomic root and slot identity remains unchanged.

I ran the focused package repeatedly under the race detector, the fork-graph and stages race suites, and all cl/beacon packages under the race detector; all passed. CI is green.

**Comment by awskii** [comment]
@domiwei addressed in c16e878f76d, both packages green under `-race`.

- Interleavings: `TestWriterWaitsForHeadStateView`, `TestWriterWaitsForPreviousHeadStateView`, `TestHeadIdentityMatchesPublishedState`.
- Prior root: `TestPreviousHeadFastPathUsesAuthoritativePriorRoot`; fails if `copy.go:40` is deleted.
- Failure identity: `TestFailedHeadUpdateKeepsHead` now asserts `StateHead()` root and slot unchanged.
